[MQTT] Setting QOS results in an error

mqtt.conf
```
openhabian.url=tcp://127.0.0.1:1883
openhabian.clientId=OpenHAB
openhabian.user=username
openhabian.pwd=***
openhabian.qos=1
openhabian.retain=true
openhabian.async=true
```

Log
```
2018-04-14 19:31:19.650 [ERROR] [org.apache.felix.configadmin        ] - [org.openhab.io.transport.mqtt.MqttService, org.osgi.service.cm.ManagedService, id=313, bundle=213/mvn:org.openhab.io/org.openhab.io.transport.mqtt/1.11.0]: Unexpected problem updating configuration org.openhab.mqtt
java.lang.NumberFormatException: For input string: "<qos>"
        at java.lang.NumberFormatException.forInputString(NumberFormatException.java:65) ~[?:?]
        at java.lang.Integer.parseInt(Integer.java:580) [?:?]
        at java.lang.Integer.parseInt(Integer.java:615) [?:?]
        at org.openhab.io.transport.mqtt.MqttService.updated(MqttService.java:88) [213:org.openhab.io.transport.mqtt:1.11.0]
        at org.apache.felix.cm.impl.helper.ManagedServiceTracker.updated(ManagedServiceTracker.java:189) [7:org.apache.felix.configadmin:1.8.16]
        at org.apache.felix.cm.impl.helper.ManagedServiceTracker.updateService(ManagedServiceTracker.java:152) [7:org.apache.felix.configadmin:1.8.16]
        at org.apache.felix.cm.impl.helper.ManagedServiceTracker.provideConfiguration(ManagedServiceTracker.java:85) [7:org.apache.felix.configadmin:1.8.16]
        at org.apache.felix.cm.impl.ConfigurationManager$ManagedServiceUpdate.provide(ConfigurationManager.java:1479) [7:org.apache.felix.configadmin:1.8.16]
        at org.apache.felix.cm.impl.ConfigurationManager$ManagedServiceUpdate.run(ConfigurationManager.java:1435) [7:org.apache.felix.configadmin:1.8.16]
        at org.apache.felix.cm.impl.UpdateThread.run0(UpdateThread.java:141) [7:org.apache.felix.configadmin:1.8.16]
        at org.apache.felix.cm.impl.UpdateThread.run(UpdateThread.java:109) [7:org.apache.felix.configadmin:1.8.16]
```

--- a/bundles/io/org.openhab.io.transport.mqtt/src/main/java/org/openhab/io/transport/mqtt/MqttService.java
+++ b/bundles/io/org.openhab.io.transport.mqtt/src/main/java/org/openhab/io/transport/mqtt/MqttService.java
@@ -1,5 +1,5 @@
 /**
- * Copyright (c) 2010-2016 by the respective copyright holders.
+ * Copyright (c) 2010-2018 by the respective copyright holders.
  *
  * All rights reserved. This program and the accompanying materials
  * are made available under the terms of the Eclipse Public License v1.0
@@ -30,16 +30,12 @@ import org.slf4j.LoggerFactory;
  * @since 1.3.0
  */
 public class MqttService implements ManagedService {
-
-    private static Logger logger = LoggerFactory.getLogger(MqttService.class);
-
     private ConcurrentHashMap<String, MqttBrokerConnection> brokerConnections = new ConcurrentHashMap<String, MqttBrokerConnection>();
-
     private EventPublisher eventPublisher;
+    private Logger logger = LoggerFactory.getLogger(MqttService.class);
 
     @Override
     public void updated(Dictionary<String, ?> properties) throws ConfigurationException {
-
         // load broker configurations from configuration file
         if (properties == null || properties.isEmpty()) {
             return;
@@ -47,11 +43,10 @@ public class MqttService implements ManagedService {
 
         Enumeration<String> keys = properties.keys();
         while (keys.hasMoreElements()) {
-
             String key = keys.nextElement();
 
             if (key.equals("service.pid")) {
-                // ignore the only non-broker property..
+                // ignore the only non-broker property
                 continue;
             }
 
@@ -66,10 +61,10 @@ public class MqttService implements ManagedService {
             String property = subkeys[1];
 
             if (StringUtils.isBlank(value)) {
-                logger.trace("Property is empty: {}", key);
+                logger.trace("Property '{}' is empty", key);
                 continue;
             } else {
-                logger.trace("Processing property: {} = {}", key, value);
+                logger.trace("Processing property '{}' = {}", key, value);
             }
 
             MqttBrokerConnection conn = brokerConnections.get(name);
@@ -85,11 +80,18 @@ public class MqttService implements ManagedService {
             } else if (property.equals("pwd")) {
                 conn.setPassword(value);
             } else if (property.equals("qos")) {
-                conn.setQos(Integer.parseInt(value));
+                int parsedQos = safeParseInt(property, value);
+                if (parsedQos > -1) {
+                    conn.setQos(parsedQos);
+                }
             } else if (property.equals("retain")) {
-                conn.setRetain(Boolean.parseBoolean(value));
+                if (validateBooleanSetting(property, value)) {
+                    conn.setRetain(Boolean.parseBoolean(value));
+                }
             } else if (property.equals("async")) {
-                conn.setAsync(Boolean.parseBoolean(value));
+                if (validateBooleanSetting(property, value)) {
+                    conn.setAsync(Boolean.parseBoolean(value));
+                }
             } else if (property.equals("clientId")) {
                 if (value.length() > 65535) {
                     logger.warn("clientId must be less than 65536 characters long");
@@ -97,13 +99,18 @@ public class MqttService implements ManagedService {
                 }
                 conn.setClientId(value);
             } else if (property.equals("allowLongerClientIds")) {
-                conn.setAllowLongerClientIds(Boolean.parseBoolean(value));
+                if (validateBooleanSetting(property, value)) {
+                    conn.setAllowLongerClientIds(Boolean.parseBoolean(value));
+                }
             } else if (property.equals("lwt")) {
                 MqttWillAndTestament will = MqttWillAndTestament.fromString(value);
                 logger.debug("Setting last will: {}", will);
                 conn.setLastWill(will);
             } else if (property.equals("keepAlive")) {
-                conn.setKeepAliveInterval(Integer.parseInt(value));
+                int parsedKeepAlive = safeParseInt(property, value);
+                if (parsedKeepAlive > -1) {
+                    conn.setKeepAliveInterval(parsedKeepAlive);
+                }
             } else {
                 logger.warn("Unrecognized property: {}", key);
             }
@@ -221,5 +228,28 @@ public class MqttService implements ManagedService {
      */
     public void unsetEventPublisher(EventPublisher eventPublisher) {
         this.eventPublisher = null;
+    }
+
+    // validates that an incoming string value from the config file represents
+    // a valid boolean value. Logs a warning if it does not.
+    private boolean validateBooleanSetting(String settingName, String value) {
+        if (value.toLowerCase().equals("true") || value.toLowerCase().equals("false")) {
+            return true;
+        }
+
+        logger.warn("Ignored invalid value for property '{}' ('{}').", settingName, value);
+        return false;
+    }
+
+    // attempts to extract an integer value from the String value.
+    // Logs a warning if parse fails.
+    private int safeParseInt(String settingName, String value) {
+        try {
+            int number = Integer.parseInt(value);
+            return number;
+        } catch (NumberFormatException e) {
+            logger.warn("Ignored invalid value for property '{}' ('{}')", settingName, value);
+            return -1;
+        }
     }
 }

